### ISSUE TITLE
no-useless-predicate should note index signatures to avoid false positives

### ISSUE THREAD

**Issue by CSchulz** [fimbullinter/wotan#522]
The rule [no-useless-predicate](https://github.com/fimbullinter/wotan/blob/master/packages/mimir/docs/no-useless-predicate.md) complains because the key is always part of the given index signature.

It should not complain if the key is optional in the object.

I think for the first version the rule should ignore assignments to variables and just honor control conditions.

```
public ngOnChanges(changes: SimpleChanges) {
    // no-useless-predicate: Expression is always truthy.
    if (changes.messages) {

export interface SimpleChanges {
    [propName: string]: SimpleChange;
}
```

**Good code:**
```
public ngOnChanges(changes: SimpleChanges) {
    if (changes.messages) {
```

**Bad code:**
```
public ngOnChanges(changes: SimpleChanges) {
    // no-useless-predicate: Expression is always truthy.
    const asd = changes.messages;
```

**Comment by ajafff** [comment]
This is a known limitation which is surprisingly not mentioned in the docs.

Detecting index signatures on property access and element access is possible. Marking as `bug` to implement just that.
If you assign the value to a variable before checking it in a condition, there's no (easy) way to detect that it came from an index signature.

```ts
declare let foo: Record<string, string>;

if (foo.bar) {} // valid
if (foo['bar']) {} // valid

const {bar} = foo;
if (bar) {} // still considered invalid
```
